src/numerics/petsc_matrix.C

```diff
@@ -1266,9 +1267,18 @@ void PetscMatrix<T>::matrix_matrix_mult (SparseMatrix<T> & X_in, SparseMatrix<T>
 
   semiparallel_only();
 
-  ierr = MatMatMult(_mat, X->_mat, MAT_INITIAL_MATRIX, PETSC_DEFAULT, &Y->_mat);
-  LIBMESH_CHKERR(ierr);
-
+  if (reuse)
+  {
+    ierr = MatMatMult(_mat, X->_mat, MAT_REUSE_MATRIX, PETSC_DEFAULT, &Y->_mat);
+    LIBMESH_CHKERR(ierr);
+  }
+  else
+  {
+    ierr = MatDestroy(&Y->_mat);
```

Comment:
This I'm not so sure about... the `Y` that is passed in could have been used for anything previously, and, for instance, it could have `_destroy_mat_on_exit == false`, in which case we should not call `MatDestroy` like this. I think it should be OK to call `Y->clear()`, which handles this distinction as well as doing some error checking/setting of flags.

Reply:
Agreed.  But also: was this a bug before?  Our existing matrix_matrix_mult() would have given us some error or some memory leak if Y had a Mat which had already been initialized?

Reply:
Thanks. This is a great suggestion.

I think this is a bug but was somehow well handled by PETSc for most cases. The memory issue came up when I was trying to do matrix inversion using `MATSOLVERSUPERLU_DIST`. The original `matrix_matrix_mult()` results in a hanging pointer and `Petsc_Superlu_dist_keyval_Delete_Fn()` was complaining about this upon existing.